src/shared/sleep-config.c

```diff
@@ -526,6 +536,88 @@ int get_total_suspend_interval(Hashmap *last_capacity, usec_t *ret) {
 
+static int read_power_supply_battery_alarm_file(const char *alarm_filepath, int *ret) {
+        _cleanup_free_ char *batterytrippoint = NULL;
+        int battery_alarm, r;
+
+        r = read_one_line_file(alarm_filepath, &batterytrippoint);
+        if (r < 0)
+                return r;
+
+        r = safe_atoi(batterytrippoint, &battery_alarm);
+        if (r < 0)
+                return r;
+
+        *ret = battery_alarm;
+
+        return 0;
+}
+
+bool battery_trip_point_alarm_exists(void) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *dev;
+        int r;
+
+        r = battery_enumerator_new(&e);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to initialize battery enumerator: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                const char *battery_name, *alarm_filepath;
+                int battery_alarm;
+
+                r = sd_device_get_property_value(dev, "POWER_SUPPLY_NAME", &battery_name);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name, ignoring: %m");
+                        continue;
+                }
+
+                alarm_filepath = path_join(POWER_SUPPLY_DIRPATH, battery_name, "alarm");
+                r = read_power_supply_battery_alarm_file(alarm_filepath, &battery_alarm);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery alarm from %s, ignoring: %m", alarm_filepath);
+                        continue;
+                }
+                if (battery_alarm <= 0)
+                        return false;
+        }
+
+        return true;
+}
+
+/* Get wakeup-type from dmi tables */
+int get_dmi_wakeup_type(char *ret) {
+        _cleanup_free_ char *s = NULL;
+        char offset_char;
+        size_t readsize;
+        int offset_int, r;
+
+        /* implementation via dmi/entries */
+        r = read_full_virtual_file(SYS_ENTRY_RAW_FILE_TYPE1, &s, &readsize);
+        if (r < 0) {
+                log_debug_errno(r, "Unable to read %s, ignoring: %m", SYS_ENTRY_RAW_FILE_TYPE1);
+                return SMBIOS_WAKEUP_BIT_UNKNOWN;
+        }
+        if (readsize < 25 || s[1] < 25) {
+                log_debug("Only read %zu bytes from %s (expected 24)", readsize, SYS_ENTRY_RAW_FILE_TYPE1);
+                return SMBIOS_WAKEUP_BIT_UNKNOWN;
+        }
+
+        offset_char = s[24];
+        if (offset_char) {
+                log_debug("DMI BIOS System Information indicates wakeup type bit is set.");
+
+                r = safe_atoi(&offset_char, &offset_int);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to parse wakeup type offset value: %m");
+
+                *ret = offset_int;
+                return SMBIOS_WAKEUP_BIT_SET;
+        }
+        log_debug("DMI BIOS System Information does not indicate wakeup type bit is set.");
+        return SMBIOS_WAKEUP_BIT_UNSET;
```

Comment:
This can be simplified: we are not interest in the exact bit set, only if it was timer or not. So instead of the enum, simply make this function return 1 if it was timer, 0 if it wasn't, and negative error at usual on errors.

Reply:
In systems like Ubuntu 22.04 which have btp support so alarm file exists but the architecture does not expose smbios dmi entries. so this will need to return 1 if wakeup bit is set no matter what. if bit  corresponds to APM, timer then hibernate

Reply:
@poettering do you know what's the story with these dmi entries not being available?